compiler/rustc_middle/messages.ftl

```diff
@@ -43,3 +43,189 @@ middle_requires_lang_item = requires `{$name}` lang_item
+    writing to {$allocation} which is read-only
+
+middle_interpret_ub_deref_function_pointer =
+    accessing {$allocation} which contains a function
+
+middle_interpret_ub_deref_vtable_pointer =
+    accessing {$allocation} which contains a vtable
+
+middle_interpret_ub_invalid_bool =
+    interpreting an invalid 8-bit value as a bool: 0x{$val}
+
+middle_interpret_ub_invalid_char =
+    interpreting an invalid 32-bit value as a char: 0x{$val}
+
+middle_interpret_ub_invalid_tag =
+    enum value has invalid tag: {$val}
+
+middle_interpret_ub_invalid_function_pointer =
+    using {$pointer} as function pointer but it does not point to a function
+
+middle_interpret_ub_invalid_vtable_pointer =
+    using {$pointer} as vtable pointer but it does not point to a vtable
+
+middle_interpret_ub_invalid_str =
+    this string is not valid UTF-8: {$err}
+
+middle_interpret_ub_invalid_uninit_bytes =
+    reading memory at {$alloc}{$access}, but memory is uninitialized at {$uninit}, and this operation requires initialized memory
+
+middle_interpret_ub_invalid_uninit_bytes_unknown =
+    using uninitialized data, but this operation requires initialized memory
+
+middle_interpret_ub_dead_local =
+    accessing a dead local variable
+
+middle_interpret_ub_scalar_size_mismatch =
+    scalar size mismatch: expected {$target_size} bytes but got {$data_size} bytes instead
+
+middle_interpret_ub_uninhabited_enum_variant_written =
+    writing discriminant of an uninhabited enum
+
+middle_partial_pointer_overwrite =
+    unable to overwrite parts of a pointer in memory at {$ptr}
+
+middle_partial_pointer_copy =
+    unable to copy parts of a pointer from memory at {$ptr}
+
+middle_ptr_as_bytes_1 =
+    this code performed an operation that depends on the underlying bytes representing a pointer
+
+middle_ptr_as_bytes_2 =
+    the absolute address of a pointer is not known at compile-time, so such operations are not supported
+
+middle_read_pointer_as_bytes =
+    unable to turn pointer into raw bytes
+
+middle_thread_local_static =
+    cannot access thread local static ({$did})
+
+middle_read_extern_static =
+    cannot read from extern static ({$did})
+
+middle_stack_frame_limit_reached =
+    reached the configured maximum number of stack frames
+
+middle_step_limit_reached =
+    exceeded interpreter step limit (see `#[const_eval_limit]`)
+
+middle_memory_exhausted =
+    tried to allocate more memory than available to compiler
+
+middle_address_space_full =
+    there are no more free addresses in the address space
+
+middle_too_generic =
+    encountered overly generic constant
+
+middle_already_reported =
+    an error has already been reported elsewhere (this should not usually be printed)
+
+middle_size_of_unsized =
+    size_of called on unsized type `{$ty}`
+
+middle_adjust_for_foreign_abi_error =
+    target architecture {$arch} does not support `extern {$abi}` ABI
+
+middle_uninit_unsized_local =
+    unsized local is used while uninitialized
```

Comment:
Is there a way to split the interpreter messages into their own file? Having everything in rustc_middle collected into a single file makes things rather messy; these things have little to do with each other apart from the fact that due to how the query system works they need to be in rustc_middle.

(Ideally all the interpreter stuff would be in rustc_const_eval but at least with the old error structure that wasn't possible. Maybe it is possible now, with the translations introducing an extra indirection?)

Reply:
I don't think it is possible without using trait objects. The procedural macro compiles the fluent resource and generate consts that are `DiagnosticMessage`s that are then used in the `derive` macro for `IntoDiagnostic`. Therefore the crate can only refer to messages that exist in their own crate

Reply:
Is it possible to move the entire function with the big `match` on `InterpError` that converts it to fluent data structures could be moved into rustc_const_eval? The type needs to be in rustc_middle but I believe this function could be moved. This requires an extension trait (since we can't use an inherent impl cross-crate) so we didn't consider this worth it in the past, but with all this fluent boilerplate I think the scale tips towards having this in rustc_const_eval.

Reply:
Moved.